common/src/main/java/me/lucko/luckperms/common/messaging/rabbitmq/RabbitMQMessenger.java

```diff
@@ -0,0 +1,185 @@
+        }
+    }
+
+    @Override
+    public void close() {
+        try {
+            this.channel.close();
+            this.connection.close();
+            this.sub.isClosed = true;
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    private static class Subscription implements Runnable {
+        private final RabbitMQMessenger parent;
+        private boolean isClosed = false;
+        private boolean firstStartup = true;
+
+        private Subscription(RabbitMQMessenger parent) {
+            this.parent = parent;
+        }
+
+        @Override
+        public void run() {
+            while (!Thread.interrupted() && !this.isClosed) {
+                try {
+                    if (!checkAndReopenConnection()) {
+                        // Sleep for 5 seconds to prevent massive spam in console
+                        Thread.sleep(5000);
+                        continue;
+                    }
+
+                    // Check connection life every every 30 seconds
+                    Thread.sleep(30_000);
+                } catch (InterruptedException ie) {
+                    Thread.currentThread().interrupt();
+                } finally {
+                    this.firstStartup = false;
+                }
+            }
+        }
+
+        private boolean checkAndReopenConnection() {
+            boolean channelIsDead = this.parent.channel == null || !this.parent.channel.isOpen();
+            if (channelIsDead) {
+                boolean connectionIsDead = this.parent.connection == null || !this.parent.connection.isOpen();
+                if (connectionIsDead) {
+                    if (!this.firstStartup) {
+                        this.parent.plugin.getLogger().warn("RabbitMQ pubsub connection dropped, trying to re-open the connection");
+                    }
+                    try {
+                        this.parent.connection = this.parent.connectionFactory.newConnection();
+                        this.parent.channel = this.parent.connection.createChannel();
+
+                        String queue = this.parent.channel.queueDeclare("", false, true, true, null).getQueue();
+                        this.parent.channel.exchangeDeclare(EXCHANGE, BuiltinExchangeType.TOPIC, false, true, null);
+                        this.parent.channel.queueBind(queue, EXCHANGE, ROUTING_KEY);
+                        this.parent.channel.basicConsume(queue, true, new ChannelListener(), (consumerTag) -> { });
+
+                        if (!this.firstStartup) {
+                            this.parent.plugin.getLogger().info("RabbitMQ pubsub connection re-established");
+                        }
+                        return true;
+                    } catch (Exception ignored) {
+                        return false;
+                    }
+                } else {
+                    try {
+                        this.parent.channel = this.parent.connection.createChannel();
```

Comment:
@lucko Oops, after that things should be declared like above
```java
String queue = this.parent.channel.queueDeclare("", false, true, true, null).getQueue();
this.parent.channel.exchangeDeclare(EXCHANGE, BuiltinExchangeType.TOPIC, false, true, null);
this.parent.channel.queueBind(queue, EXCHANGE, ROUTING_KEY);
this.parent.channel.basicConsume(queue, true, new ChannelListener(), (consumerTag) -> { });
```

Reply:
Ah.

I mean, if the connection isn't "dead" - does it need to be reestablished?

Reply:
It works like this: `connection -> channel -> listener (exchange/queue/consumer)`.

If the channel or connection was closed, then the exchange queue and consumer (listener) will be closed and must be re-created in both cases, if the connection was closed then new channel must be created.